Vic2ToHoI4/Source/HOI4World/WarCreator/MapUtils.cpp

```diff
@@ -0,0 +1,340 @@
+{
+	const auto countriesAndDistances = distancesBetweenCountries.find(country);
+	if (countriesAndDistances == distancesBetweenCountries.end())
+	{
+		return {};
+	}
+
+	std::set<std::string> nearbyCountries;
+	for (const auto& [tag, distance]: countriesAndDistances->second)
+	{
+		if (distance <= range)
+		{
+			nearbyCountries.insert(tag);
+		}
+	}
+
+	return nearbyCountries;
+}
+
+
+std::set<std::string> HoI4::MapUtils::getFarCountries(const std::string& country, float range) const
+{
+	const auto countriesAndDistances = distancesBetweenCountries.find(country);
+	if (countriesAndDistances == distancesBetweenCountries.end())
+	{
+		return {};
+	}
+
+	std::set<std::string> nearbyCountries;
+	for (const auto& [tag, distance]: countriesAndDistances->second)
+	{
+		if (distance > range)
+		{
+			nearbyCountries.insert(tag);
+		}
+	}
+
+	return nearbyCountries;
+}
+
+
+std::vector<std::string> HoI4::MapUtils::getGPsByDistance(const Country& country,
+	 const std::vector<std::shared_ptr<Country>>& greatPowers) const
+{
+	std::map<float, std::shared_ptr<Country>> distanceToGPMap;
+	for (const auto& greatPower: greatPowers)
+	{
+		auto distance = getDistanceBetweenCapitals(country, *greatPower);
+		if (distance)
+		{
+			distanceToGPMap.insert(std::make_pair(*distance, greatPower));
+		}
+	}
+
+	std::vector<std::string> orderedGreatPowers(distanceToGPMap.size());
+	for (const auto& [unused, country]: distanceToGPMap)
+	{
+		orderedGreatPowers.push_back(country->getTag());
+	}
+
+	return orderedGreatPowers;
+}
+
+
+std::optional<HoI4::Coordinate> HoI4::MapUtils::getProvincePosition(int provinceNum) const
+{
+	if (const auto province = provincePositions.find(provinceNum); province != provincePositions.end())
+	{
+		return province->second;
+	}
+
+	return std::nullopt;
+}
+
+
+float HoI4::MapUtils::getDistanceSquaredBetweenPoints(const Coordinate& point1, const Coordinate& point2) const
+{
+	auto xDistance = static_cast<float>(abs(point2.x - point1.x));
+	if (xDistance > halfMapWidth)
+	{
+		xDistance = mapWidth - xDistance;
+	}
+
+	const auto yDistance = static_cast<float>(point2.y - point1.y);
+
+	return xDistance * xDistance + yDistance * yDistance;
+}
+
+
+std::optional<float> HoI4::MapUtils::getDistanceBetweenCountries(const Country& country1, const Country& country2) const
+{
+	auto distanceBetweenCapitals = getDistanceBetweenCapitals(country1, country2);
+	if (!distanceBetweenCapitals)
+	{
+		return std::nullopt;
+	}
+
+	auto distanceSquared = *distanceBetweenCapitals * *distanceBetweenCapitals;
+	for (auto province1: country1.getProvinces())
+	{
+		auto province1Position = getProvincePosition(province1);
+		if (!province1Position)
+		{
+			continue;
+		}
+
+		for (auto province2: country2.getProvinces())
+		{
+			auto province2Position = getProvincePosition(province2);
+			if (!province2Position)
+			{
+				continue;
+			}
+
+			distanceSquared =
```

Comment:
This is insanely expensive and should be done once per entire converter session for all provinces. Please make sure you're never doing it multiple times without caching.

Reply:
Cached now.